### PR TITLE
rngd: Start early during boot and encrypted swap entropy fix

### LINKED ISSUE #71204: [19.09] swapDevices.*.randomEncryption possible entropy starvation during boot
**Describe the bug**
Since upgrading to 19.09, my laptop's boot process is delayed by about one minute due to entropy starvation during the boot process

**To Reproduce**
1. swap configuration:
 { device = "/dev/disk/by-id/yadayada-part2"; randomEncryption.enable = true; randomEncryption.source = "/dev/random"; }
2. [possibly configure other services that also need entropy during boot]
3. reboot

**Expected behavior**
A quick boot as in 19.03 with max 1 second used by the swap initalization.

**Screenshots**
$ systemd-analyze plot | fgrep mkswap
  <text class="left" x="1018.920" y="694.000">mkswap-dev-disk-byx2did-nvmex2dSamsung_SSD_970_EVO_2TB_S464NB0M401994Dx2dpart2.service (1min 723ms)</text>

**Additional context**
This is on physical hardware, not on VM. Replacing random with urandom leads to instant boot, but has the obvious problem with the missing entropy.

**Metadata**
# nix run nixpkgs.nix-info -c nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.79, NixOS, 19.09.830.496bdc7e334 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3`
 - channels(root): `"nixos-19.09.830.496bdc7e334"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
swapDevices.*.randomEncryption
# a list of nixos modules affected by the problem
module: nixos/modules/config/swap.nix
```

### PR DESCRIPTION
**Pull request by tokudan**
###### Motivation for this change

Intended to fix entropy starvation during boot (fixes #71204)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @abbradar

### CODE REVIEW

File: nixos/modules/security/rngd.nix [hunk 1]
```diff
@@ -39,12 +39,12 @@ in
 
       description = "Hardware RNG Entropy Gatherer Daemon";
 
+      unitConfig.DefaultDependencies = false;
       serviceConfig = {
         ExecStart = "${pkgs.rng-tools}/sbin/rngd -f"
           + optionalString cfg.debug " -d";
         NoNewPrivileges = true;
         PrivateNetwork = true;
-        PrivateTmp = true;
```
Comment (grahamc) [inline nixos/modules/security/rngd.nix]: why the removal?

Comment (tokudan) [reply nixos/modules/security/rngd.nix]: Added the explanation in the code.

**Comment by abbradar** [comment]
Interesting, thanks! I'll be able to review this in the upcoming week.

(Sorry for the absence, had a fun traveling accident which greatly prolonged my vacation :D)

**Comment by tokudan** [comment]
Squashed a commit and rebased onto the current master

**Comment by tokudan** [comment]
I believe this is ready for review/merging

File: nixos/modules/config/swap.nix [hunk 2]
```diff
@@ -185,6 +185,7 @@ in
           { description = "Initialisation of swap device ${sw.device}";
             wantedBy = [ "${realDevice'}.swap" ];
             before = [ "${realDevice'}.swap" ];
+            after = mkIf (config.security.rngd.enable && sw.randomEncryption.enable) [ "rngd.service" ];
```
Comment (rnhmjoj) [inline nixos/modules/config/swap.nix]: I think a comment mentioning the entropy starvation here would also be helpful.

Comment (tokudan) [reply nixos/modules/config/swap.nix]: Added some more comments to my commits

**Review by rnhmjoj — APPROVED** [review]
Looks good to me.

**Pull request merged**